openff/bespokefit/executor/services/optimizer/worker.py

```diff
@@ -26,29 +29,35 @@ def optimize(self, optimization_input_json: str) -> str:
     input_schema = parse_raw_as(
         Union[BespokeOptimizationSchema], optimization_input_json
     )
-    input_schema.id = self.request.id
-
+    input_schema.id = self.request.id or input_schema.id
     input_force_field = ForceField(input_schema.initial_force_field)
 
     stage_results = []
+    # run all optimisations in the task folder
+    os.makedirs(input_schema.id, exist_ok=True)
+    with temporary_cd(input_schema.id):
+        for i, stage in enumerate(input_schema.stages):
 
-    for stage in input_schema.stages:
-
-        optimizer = get_optimizer(stage.optimizer.type)
-        result = optimizer.optimize(
-            stage, input_force_field, keep_files=settings.BEFLOW_KEEP_FILES
-        )
+            optimizer = get_optimizer(stage.optimizer.type)
+            result = optimizer.optimize(
+                schema=stage,
+                initial_force_field=input_force_field,
+                keep_files=settings.BEFLOW_KEEP_FILES,
+                root_directory=f"stage_{i}",
+            )
 
-        stage_results.append(result)
+            stage_results.append(result)
 
-        if result.status != "success":
-            break
+            if result.status != "success":
+                break
 
-        input_force_field = ForceField(
-            ForceField(
-                result.refit_force_field, allow_cosmetic_attributes=True
-            ).to_string(discard_cosmetic_attributes=True)
-        )
+            input_force_field = ForceField(
+                ForceField(
+                    result.refit_force_field, allow_cosmetic_attributes=True
+                ).to_string(discard_cosmetic_attributes=True)
+            )
+    if not settings.BEFLOW_KEEP_FILES:
+        shutil.rmtree(input_schema.id, ignore_errors=True)
```

Comment:
Hmm we may need to be a bit careful here - tasks should be idempotent, which basically means you should be able to run the same task twice and not have it interfere with itself.

To ensure this is the case, we should maybe run the optimization in a temp directory, and then copy it back to a non-temp directory when it's finished if keep files is true?